File: backend/app/services/report_generator.py
import os
import io
import csv
import logging
from typing import Dict, Any, Tuple
from motor.motor_asyncio import AsyncIOMotorDatabase
from bson import ObjectId

from reportlab.lib.pagesizes import letter
from reportlab.platypus import SimpleDocTemplate, Paragraph, Spacer, Table, TableStyle, HRFlowable
from reportlab.lib.styles import getSampleStyleSheet, ParagraphStyle
from reportlab.lib import colors

logger = logging.getLogger("app.services.report_generator")

async def generate_assessment_report(
    db: AsyncIOMotorDatabase,
    assessment_id: str,
    export_format: str = "pdf"
) -> Tuple[bytes, str, str]:
    """
    Generates a structured PDF or CSV report for an assessment.
    Returns tuple: (file_bytes, filename, media_type)
    """
    # Fetch assessment, inputs, recommendations, facility
    assessment = await db.assessments.find_one({"_id": assessment_id})
    if not assessment and ObjectId.is_valid(assessment_id):
        assessment = await db.assessments.find_one({"_id": ObjectId(assessment_id)})

    if not assessment:
        raise ValueError("Assessment not found")

    facility_id = assessment.get("facility_id")
    facility = await db.facilities.find_one({"_id": facility_id})
    if not facility and ObjectId.is_valid(facility_id):
        facility = await db.facilities.find_one({"_id": ObjectId(facility_id)})

    facility_name = facility.get("name", "Industrial Facility") if facility else "Industrial Facility"
    industry = facility.get("industry", "N/A") if facility else "N/A"

    cursor_inputs = db.process_inputs.find({"assessment_id": assessment_id})
    inputs = await cursor_inputs.to_list(length=500)

    cursor_recs = db.recommendations.find({"assessment_id": assessment_id})
    recs = await cursor_recs.to_list(length=100)

    total_co2e = float(assessment.get("total_co2e", 0.0))

    if export_format.lower() == "csv":
        output = io.StringIO()
        writer = csv.writer(output)

        writer.writerow(["ASSESSMENT EMISSIONS & CIRCULAR RECOMMENDATION REPORT"])
        writer.writerow(["Facility Name", facility_name])
        writer.writerow(["Industry", industry])
        writer.writerow(["Assessment ID", assessment_id])
        writer.writerow(["Total Calculated CO2e (tonnes)", total_co2e])
        writer.writerow([])

        writer.writerow(["PROCESS INPUT LINE ITEMS"])
        writer.writerow(["Category", "Subtype", "Quantity", "Unit", "Computed CO2e (tonnes)"])
        for inp in inputs:
            writer.writerow([
                inp.get("category"),
                inp.get("subtype"),
                inp.get("quantity"),
                inp.get("unit"),
                inp.get("computed_co2e", 0.0)
            ])
        writer.writerow([])

        writer.writerow(["RECOMMENDED CIRCULAR INTERVENTIONS"])
        writer.writerow(["Intervention Name", "Category", "Score", "Score Source", "Cost Range ($)", "CO2 Reduction Range (%)", "Applicable Leak Point"])
        for r in recs:
            intv = r.get("intervention", {})
            c_range = r.get("estimated_cost_range", [0, 0])
            r_range = r.get("estimated_co2_reduction_range", [0, 0])
            writer.writerow([
                intv.get("name"),
                intv.get("category"),
                r.get("score"),
                r.get("score_source"),
                f"${c_range[0]} - ${c_range[1]}",
                f"{r_range[0]}% - {r_range[1]}%",
                r.get("applicable_leak_point")
            ])

        csv_bytes = output.getvalue().encode("utf-8")
        filename = f"assessment_report_{assessment_id}.csv"
        return csv_bytes, filename, "text/csv"

    # Default: PDF format using ReportLab
    pdf_buffer = io.BytesIO()
    doc = SimpleDocTemplate(pdf_buffer, pagesize=letter, rightMargin=36, leftMargin=36, topMargin=36, bottomMargin=36)
    styles = getSampleStyleSheet()

    title_style = ParagraphStyle(
        'DocTitle',
        parent=styles['Heading1'],
        fontName='Helvetica-Bold',
        fontSize=20,
        textColor=colors.HexColor('#0F172A'),
        spaceAfter=6
    )
    subtitle_style = ParagraphStyle(
        'DocSubTitle',
        parent=styles['Normal'],
        fontName='Helvetica',
        fontSize=10,
        textColor=colors.HexColor('#64748B'),
        spaceAfter=15
    )
    heading2_style = ParagraphStyle(
        'Heading2',
        parent=styles['Heading2'],
        fontName='Helvetica-Bold',
        fontSize=14,
        textColor=colors.HexColor('#1E293B'),
        spaceBefore=12,
        spaceAfter=8
    )

    story = []
    story.append(Paragraph("Industrial Emission Leak-Point & Circular Action Report", title_style))
    story.append(Paragraph(f"Facility: <b>{facility_name}</b> | Industry: <b>{industry.capitalize()}</b> | Date: {assessment.get('created_at', '2026-09-12')}", subtitle_style))
    story.append(HRFlowable(width="100%", thickness=1.5, color=colors.HexColor('#2563EB'), spaceAfter=15))

    # Summary KPI table
    kpi_data = [
        ["Total CO2e Footprint", "Status", "Total Process Inputs", "Recommendations"],
        [f"{total_co2e:.2f} t CO2e", assessment.get("status", "complete").upper(), str(len(inputs)), str(len(recs))]
    ]
    kpi_table = Table(kpi_data, colWidths=[130, 130, 130, 130])
    kpi_table.setStyle(TableStyle([
        ('BACKGROUND', (0,0), (-1,0), colors.HexColor('#F1F5F9')),
        ('TEXTCOLOR', (0,0), (-1,0), colors.HexColor('#334155')),
        ('FONTNAME', (0,0), (-1,0), 'Helvetica-Bold'),
        ('FONTSIZE', (0,0), (-1,-1), 10),
        ('ALIGN', (0,0), (-1,-1), 'CENTER'),
        ('BOTTOMPADDING', (0,0), (-1,-1), 8),
        ('TOPPADDING', (0,0), (-1,-1), 8),
        ('GRID', (0,0), (-1,-1), 0.5, colors.HexColor('#CBD5E1')),
    ]))
    story.append(kpi_table)
    story.append(Spacer(1, 15))

    # Process Inputs Table
    story.append(Paragraph("1. Process Activity & Emission Breakdown (Calculated)", heading2_style))
    input_table_data = [["Category", "Subtype / Material", "Quantity", "Unit", "Computed CO2e (t)"]]
    for inp in inputs:
        input_table_data.append([
            inp.get("category", "").capitalize(),
            inp.get("subtype", "").replace("_", " ").title(),
            f"{float(inp.get('quantity', 0)):,.2f}",
            inp.get("unit", ""),
            f"{float(inp.get('computed_co2e', 0.0)):.4f}"
        ])
    
    input_table = Table(input_table_data, colWidths=[90, 160, 90, 80, 100])
    input_table.setStyle(TableStyle([
        ('BACKGROUND', (0,0), (-1,0), colors.HexColor('#1E293B')),
        ('TEXTCOLOR', (0,0), (-1,0), colors.white),
        ('FONTNAME', (0,0), (-1,0), 'Helvetica-Bold'),
        ('FONTSIZE', (0,0), (-1,-1), 9),
        ('GRID', (0,0), (-1,-1), 0.5, colors.HexColor('#E2E8F0')),
        ('TOPPADDING', (0,0), (-1,-1), 6),
        ('BOTTOMPADDING', (0,0), (-1,-1), 6),
    ]))
    story.append(input_table)
    story.append(Spacer(1, 15))

    # Recommendations Table
    story.append(Paragraph("2. Ranked Circular Economy Interventions (Scored & Costed)", heading2_style))
    rec_table_data = [["Rank / Score", "Intervention Name", "Est. Investment", "CO2 Cut (%)", "Source"]]
    for r in recs[:10]:
        intv = r.get("intervention", {})
        c_range = r.get("estimated_cost_range", [0, 0])
        r_range = r.get("estimated_co2_reduction_range", [0, 0])
        rec_table_data.append([
            f"{r.get('score')}/100",
            intv.get("name", ""),
            f"${c_range[0]:,} - ${c_range[1]:,}",
            f"{r_range[0]}% - {r_range[1]}%",
            r.get("score_source", "ml").upper()
        ])
    
    rec_table = Table(rec_table_data, colWidths=[80, 190, 110, 80, 60])
    rec_table.setStyle(TableStyle([
        ('BACKGROUND', (0,0), (-1,0), colors.HexColor('#0F766E')),
        ('TEXTCOLOR', (0,0), (-1,0), colors.white),
        ('FONTNAME', (0,0), (-1,0), 'Helvetica-Bold'),
        ('FONTSIZE', (0,0), (-1,-1), 9),
        ('GRID', (0,0), (-1,-1), 0.5, colors.HexColor('#E2E8F0')),
        ('TOPPADDING', (0,0), (-1,-1), 6),
        ('BOTTOMPADDING', (0,0), (-1,-1), 6),
    ]))
    story.append(rec_table)

    doc.build(story)
    pdf_bytes = pdf_buffer.getvalue()
    filename = f"assessment_report_{assessment_id}.pdf"
    return pdf_bytes, filename, "application/pdf"

File: backend/src/services/reportGenerator.js
const { v4: uuidv4 } = require('uuid');

/**
 * Generate a CSV report string for an assessment
 */
function generateCSV(facility, assessment, summary, leakPoints, recommendations) {
  const rows = [];

  // Header
  rows.push('HACKOUT 2026 – Industrial Emission Reduction Platform');
  rows.push(`Certified Emission Report — ${new Date().toISOString()}`);
  rows.push('');

  // Facility Info
  rows.push('FACILITY INFORMATION');
  rows.push(`Name,${facility.name}`);
  rows.push(`Industry,${facility.industry}`);
  rows.push(`Size,${facility.facility_size}`);
  rows.push(`Region,${facility.region}`);
  rows.push(`Production Volume,${facility.production_volume || 'N/A'}`);
  rows.push('');

  // Assessment Overview
  rows.push('EMISSION OVERVIEW');
  rows.push(`Total CO2e (tonnes),${summary.total_co2e}`);
  rows.push(`Assessment Status,${assessment.status}`);
  rows.push(`Completed At,${assessment.completed_at || 'N/A'}`);
  rows.push(`Energy Emissions (t CO2e),${summary.category_totals?.energy || 0}`);
  rows.push(`Material Emissions (t CO2e),${summary.category_totals?.material || 0}`);
  rows.push(`Waste Emissions (t CO2e),${summary.category_totals?.waste || 0}`);
  rows.push('');

  // Line Items
  rows.push('PROCESS INPUTS & EMISSION LINE ITEMS');
  rows.push('ID,Category,Subtype,Quantity,Unit,Emission Factor,Factor Unit,CO2e (tonnes),% Contribution');
  for (const item of (summary.line_items || [])) {
    rows.push([
      item.id, item.category, item.subtype, item.quantity, item.unit,
      item.emission_factor, item.factor_unit,
      item.co2e, (item.pct_contribution * 100).toFixed(1) + '%'
    ].join(','));
  }
  rows.push('');

  // Leak Points
  rows.push('EMISSION LEAK POINTS (RANKED)');
  rows.push('Rank,Category,Subtype,CO2e (tonnes),% Contribution,Severity');
  for (const lp of (leakPoints || [])) {
    rows.push([
      lp.rank, lp.category, lp.subtype, lp.co2e,
      (lp.pct_contribution * 100).toFixed(1) + '%', lp.severity
    ].join(','));
  }
  rows.push('');

  // Recommendations
  rows.push('TOP RECOMMENDATIONS');
  rows.push('Rank,Intervention,Category,Score,CO2 Reduction (t/yr),Cost Range (USD),Payback (months),Difficulty');
  const topRecs = (recommendations || []).slice(0, 5);
  topRecs.forEach((rec, idx) => {
    rows.push([
      idx + 1,
      `"${rec.intervention?.name || rec.name}"`,
      rec.intervention?.category || rec.category,
      rec.score,
      `${rec.estimated_co2_reduction_range?.[0]}-${rec.estimated_co2_reduction_range?.[1]}`,
      `$${rec.estimated_cost_range?.[0]}-$${rec.estimated_cost_range?.[1]}`,
      rec.payback_period_months,
      rec.implementation_difficulty
    ].join(','));
  });

  rows.push('');
  rows.push(`Report Generated,${new Date().toISOString()}`);
  rows.push('Platform,HackOut 2026 — Industrial Emission Reduction Platform');

  return rows.join('\n');
}

/**
 * Generate a text-based PDF-like report (returned as text/plain when pdfkit unavailable)
 */
function generateTextReport(facility, assessment, summary, leakPoints, recommendations) {
  const sep = '═'.repeat(60);
  const lines = [];

  lines.push(sep);
  lines.push('  HACKOUT 2026 — CERTIFIED EMISSION REPORT');
  lines.push(`  Generated: ${new Date().toISOString()}`);
  lines.push(sep);
  lines.push('');

  lines.push('FACILITY');
  lines.push(`  Name:             ${facility.name}`);
  lines.push(`  Industry:         ${facility.industry}`);
  lines.push(`  Size:             ${facility.facility_size}`);
  lines.push(`  Region:           ${facility.region}`);
  lines.push('');

  lines.push('EMISSION OVERVIEW');
  lines.push(`  Total CO₂e:       ${summary.total_co2e} tonnes`);
  lines.push(`  Energy:           ${summary.category_totals?.energy || 0} t CO₂e`);
  lines.push(`  Material:         ${summary.category_totals?.material || 0} t CO₂e`);
  lines.push(`  Waste:            ${summary.category_totals?.waste || 0} t CO₂e`);
  lines.push('');

  lines.push('TOP LEAK POINTS');
  (leakPoints || []).slice(0, 5).forEach((lp) => {
    lines.push(`  ${lp.rank}. [${lp.severity.toUpperCase()}] ${lp.subtype} — ${lp.co2e} t CO₂e (${(lp.pct_contribution * 100).toFixed(1)}%)`);
  });
  lines.push('');

  lines.push('TOP RECOMMENDATIONS');
  (recommendations || []).slice(0, 5).forEach((rec, idx) => {
    lines.push(`  ${idx + 1}. ${rec.intervention?.name || rec.name} (Score: ${rec.score})`);
    lines.push(`     CO₂ Reduction: ${rec.estimated_co2_reduction_range?.[0]}-${rec.estimated_co2_reduction_range?.[1]} t/yr`);
    lines.push(`     Cost: $${rec.estimated_cost_range?.[0]}-$${rec.estimated_cost_range?.[1]} | Payback: ${rec.payback_period_months} months`);
  });
  lines.push('');
  lines.push(sep);

  return lines.join('\n');
}

module.exports = { generateCSV, generateTextReport };
